google/cloud/storage/transfer_manager.py

```diff
@@ -0,0 +1,533 @@
+
+    :raises: :exc:`concurrent.futures.TimeoutError` if deadline is exceeded.
+
+    :rtype: list
+    :returns: A list of results corresponding to, in order, each item in the
+        input list. If an exception was received, it will be the result
+        for that operation. Otherwise, the return value from the successful
+        upload method is used (typically, None).
+    """
+    if upload_kwargs is None:
+        upload_kwargs = {}
+    if skip_if_exists:
+        upload_kwargs["if_generation_match"] = 0
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
+        futures = []
+        for path_or_file, blob in file_blob_pairs:
+            method = (
+                blob.upload_from_filename
+                if isinstance(path_or_file, str)
+                else blob.upload_from_file
+            )
+            futures.append(executor.submit(method, path_or_file, **upload_kwargs))
+    results = []
+    concurrent.futures.wait(
+        futures, timeout=deadline, return_when=concurrent.futures.ALL_COMPLETED
+    )
+    for future in futures:
+        exp = future.exception()
+
+        # If raise_exception is False, don't call future.result()
+        if exp and not raise_exception:
+            results.append(exp)
+        # If skip_if_exists and the exception is PreconditionFailed, do same.
+        elif exp and skip_if_exists and isinstance(exp, exceptions.PreconditionFailed):
+            results.append(exp)
+        # Get the real result. If there was an exception not handled above,
+        # this will raise it.
+        else:
+            results.append(future.result())
+    return results
+
+
+def download_many(
+    blob_file_pairs,
+    download_kwargs=None,
+    max_workers=None,
+    deadline=None,
+    raise_exception=False,
+):
+    """Download many blobs concurrently via a worker pool.
+
+    This function is a FEATURE PREVIEW: the API may change in a future version.
+
+    :type blob_file_pairs: List(Tuple('google.cloud.storage.blob.Blob', IOBase or str))
+    :param blob_file_pairs:
+        A list of tuples of blob and a file or filename. Each blob will be
+        downloaded to the corresponding blob by using blob.download_to_file() or
+        blob.download_to_filename() as appropriate.
+
+    :type download_kwargs: dict
+    :param download_kwargs:
+        A dictionary of keyword arguments to pass to the download method. Refer
+        to the documentation for blob.download_to_file() or
+        blob.download_to_filename() for more information. The dict is directly
+        passed into the download methods and is not validated by this function.
+
+    :type max_workers: int
+    :param max_workers:
+        The number of workers (effectively, the number of threads) to use in
+        the worker pool. Refer to concurrent.futures.ThreadPoolExecutor
+        documentation for details.
+
+    :type deadline: int
+    :param deadline:
```

Comment:
Is there any guidance on what to set this value to?

Reply:
Under normal operation it should be left blank. I'll add the explanation of why to the docstrings. Thanks.